Tools/clinic/clinic.py

```diff
@@ -4311,7 +4311,7 @@ def __init__(self, clinic: Clinic) -> None:
         self.reset()
 
     def reset(self) -> None:
-        self.function = None
+        self.function: Function | None = None
```

Comment:
For attributes that aren't initialised in an `__init__` method, the usual convention is to put the attribute annotation in the class body directly above the `__init__` method, i.e.:

```diff
 class DSLParser:
+    function: Function | None
+
     def __init__(self, clinic: Clinic) -> None:
         self.clinic = clinic

@@ -4311,7 +4313,7 @@ def __init__(self, clinic: Clinic) -> None:
         self.reset()

     def reset(self) -> None:
-        self.function: Function | None = None
+        self.function = None
         self.state: StateKeeper = self.state_dsl_start
         self.parameter_indent = None
         self.keyword_only = False
```

If you follow this convention, it makes it easier for readers of the code to find all the attribute annotations for any given class. They'll all either be in the `__init__` method, or directly above it.

Reply:
Yeah, that's definitely an improvement.

Reply:
We should annotate all these attributes like that. Preferably in a separate PR.

Reply:
See gh-106357